plugin/forward/forward.go

```diff
@@ -110,7 +110,7 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	list := f.List()
 	deadline := time.Now().Add(defaultTimeout)
 	start := time.Now()
-	for time.Now().Before(deadline) {
+	for time.Now().Before(deadline) && ctx.Err() == nil {
```

Comment:
I would expect that context cancellation should be handled by `proxy.Connect` https://github.com/coredns/coredns/blob/master/plugin/forward/forward.go#L152
but currently it isn't as it seems?

IMHO the fix you propose will not correctly cancel querying when forward plugin is [retrying query over TCP](https://github.com/coredns/coredns/blob/master/plugin/forward/forward.go#L151-L163) for example

Reply:
> I would expect that context cancellation should be handled by proxy.Connect https://github.com/coredns/coredns/blob/master/plugin/forward/forward.go#L152
but currently it isn't as it seems?

It may pass context to subsequent dials and if they respect context cancelation too, these functions will return error and it's all context handling expected to be done from this side. But the outer loop explicitly waits till certain amount of time, hence that wait has to take context into account.

Reply:
Oh, now I see. proxy.Connect accepts ctx, but doesn't actually use it. Apparently that has to be addresses separately.

Reply:
> Oh, now I see. proxy.Connect accepts ctx, but doesn't actually use it. Apparently that has to be addressed separately.

agreed, can be handled separately

Reply:
We have multiple places in the loop body where we break the loop because of some errors, e.g. see lines 184, 194. Should we follow the same pattern for consistency?
With proposed implementation the error message returned by the forward plugin will be incorrect.

Reply:
Though, the error message like "no healthy proxies" may appear in existing code as well, when default deadline expires, but not all proxies were tried yet. So, this PR doesn't add new issue. Should be ok as a simple fix.